solidity/contracts/bridge/Frauds.sol

```diff
@@ -0,0 +1,577 @@
+    ) external {
+        require(
+            _newNotifierRewardMultiplier <= 100,
+            "Fraud notifier reward multiplier must be <= 100"
+        );
+        self.notifierRewardMultiplier = _newNotifierRewardMultiplier;
+        emit FraudNotifierRewardMultiplierUpdated(_newNotifierRewardMultiplier);
+    }
+
+    /// @notice Sets the new value for the `challengeDefeatTimeout` parameter.
+    /// @param _newChallengeDefeatTimeout the new value for `challengeDefeatTimeout`
+    /// @dev The value of `challengeDefeatTimeout` must be > 0.
+    function setChallengeDefeatTimeout(
+        Data storage self,
+        uint256 _newChallengeDefeatTimeout
+    ) external {
+        require(
+            _newChallengeDefeatTimeout > 0,
+            "Fraud challenge defeat timeout must be > 0"
+        );
+        self.challengeDefeatTimeout = _newChallengeDefeatTimeout;
+        emit FraudChallengeDefeatTimeoutUpdated(_newChallengeDefeatTimeout);
+    }
+
+    /// @notice Sets the new value for the `challengeDepositAmount` parameter.
+    /// @param _newChallengeDepositAmount the new value for `challengeDepositAmount`
+    /// @dev The value of `challengeDepositAmount` must be > 0.
+    function setChallengeDepositAmount(
+        Data storage self,
+        uint256 _newChallengeDepositAmount
+    ) external {
+        require(
+            _newChallengeDepositAmount > 0,
+            "Fraud challenge deposit amount must be > 0"
+        );
+        self.challengeDepositAmount = _newChallengeDepositAmount;
+        emit FraudChallengeDepositAmountUpdated(_newChallengeDepositAmount);
+    }
+
+    /// @notice Extracts the UTXO keys from the given preimage used during
+    ///         signing of a witness input.
+    /// @param preimage The preimage which produces sighash used to generate the
+    ///        ECDSA signature that is the subject of the fraud claim. It is a
+    ///        serialized subset of the transaction. The exact subset used as
+    ///        the preimage depends on the transaction input the signature is
+    ///        produced for. See BIP-143 for reference
+    /// @return utxoKey UTXO key that identifies spent input.
+    function extractUtxoKeyFromWitnessPreimage(bytes memory preimage)
+        internal
+        returns (uint256 utxoKey)
+    {
+        // The expected structure of the preimage created during signing of a
+        // witness input:
+        // - transaction version (4 bytes)
+        // - hash of previous outpoints of all inputs (32 bytes)
+        // - hash of sequences of all inputs (32 bytes)
+        // - outpoint (hash + index) of the input being signed (36 bytes)
+        // - the unlocking script of the input (variable length)
+        // - value of the outpoint (8 bytes)
+        // - sequence of the input being signed (4 bytes)
+        // - hash of all outputs (32 bytes)
+        // - transaction locktime (4 bytes)
+        // - sighash type (4 bytes)
```

Comment:
For future-us can you provide a reference to the code or documentation (just as a message in this thread) explaining this structure? Where did we get this information from? Same for non-witness preimage.

Reply:
I added the information how the preimage is constructed in the PR's description.

Reply:
Can we also add a reference in the TODO comment?